src/pymor/algorithms/rand_la.py

```diff
@@ -2,20 +2,373 @@
+
+        W = self.A.source.random(k, distribution='normal')
+        if self.complex:
+            W += 1j * self.A.source.random(k, distribution='normal')
+        AW = self.A.apply(W)
+        self._samplevecs.append(AW)
+        self._norms.extend(AW.norm(self.range_product))
+
+        if self.complement_basis is not None:
+            coeffs = self._orthonormalize(self._samplevecs[-k:], self.complement_basis)
+            self._projection_coeffs = np.vstack([self._projection_coeffs, coeffs])
+
+    def _update_projection_coeffs(self):
+        if self.complement_basis is None:
+            return
+        new_basis_vecs = len(self.complement_basis) - self._projection_coeffs.shape[1]
+        if new_basis_vecs == 0:
+            return
+        new_coeffs = self._orthonormalize(self._samplevecs, self.complement_basis[-new_basis_vecs:])
+        self._projection_coeffs = np.hstack([self._projection_coeffs, new_coeffs])
+
+    def estimate_norm(self, num_testvecs, p_fail, complement_basis_size=None):
+        r"""Randomized operator norm estimator from :cite:`BS18` (Definition 3.1).
+
+        An upper bound on the operator norm with probability greater than or equal to
+        :math:`1-\texttt{p_fail}` is given by
+
+        .. math::
+            \epsilon_{\mathrm{est}}=c_{\mathrm{est}}\cdot\max_{\omega\in\Omega}
+            \lVert A\omega\rVert_{R}
+
+        with
+
+        .. math::
+            c_{\mathrm{est}}
+            =\frac{1}{\sqrt{2\lambda_{\mathrm{min}}}\operatorname{erf}^{-1}
+            \left(\texttt{p_fail}^{1/\texttt{num_testvecs}}\right)},
+
+        where :math:`\Omega` is a set of `num_testvecs` random vectors, :math:`R` denotes the inner
+        product of the range, :math:`\lambda_min` is a lower bound on the smallest eigenvalue of
+        the inner product of the source of :math:`A`.
+
+        Parameters
+        ----------
+        num_testvecs
+            Number of test vectors used for estimation. Increasing this number will likely improve
+            the effectivity of the estimator, i.e. the sharpness of the bound. See :cite:`BS18`.
+        p_fail
+            Maximum failure probabilty of the estimate.
+
+        Returns
+        -------
+        norm
+            An approximate upper bound on the operator norm.
+        """
+        assert 0 < num_testvecs and isinstance(num_testvecs, Integral)
+        assert 0 < p_fail < 1
+
+        self._compute_lambda_min()
+        self._update_projection_coeffs()
+        self._draw_samples(num_testvecs)
+
+        norms = np.array(self._norms[:num_testvecs])
+        if self.complement_basis is not None:
+            if complement_basis_size is None:
+                complement_basis_size = len(self.complement_basis)
+            assert complement_basis_size <= len(self.complement_basis)
+            norms = np.sqrt(
+                np.abs(norms**2) - spla.norm(self._projection_coeffs[:num_testvecs, :complement_basis_size], axis=1)**2
+            )
```

Comment:
These lines look a bit fishy.. @sdrave What are you computing here?

Reply:
The norms of the images of the test vectors projected to the orthogonal complement of the current range basis. Why fishy?

Reply:
Ah sorry I misread the brackets, I thought you were squaring the `np.sqrt`. Nonetheless, I think we should add an absolute value here to avoid negative numbers inside of the root.

Reply:
OK it seems like you copied the error with the brackets from my code. Sorry about that. I noticed two things:
1. the error estimator does not go below about `3e-7`, which is why for the tolerance given, it never stops. 
2. the old version uses `atol=0` and `rtol=0` in `gram_schmidt` opposed to the default values in the new version.

Reply:
With the changes I just pushed, I now get the exact same convergence history as for main -- until the estimated error reaches half machine precision. I believe that the square root of squared differences expression here is the culprit. So the question is how to proceed here.

We introduced this new approach for estimating the error in order to be able to estimate the error for smaller basis sizes without having to recompute the projection onto the orthogonal complement.  Personally, I am not too much interested in that feature. As long as the number of test vectors is not increased, these estimates could simply be returned as the convergence history of the algorithm. It would also be possible to later increase the number of test vectors or the basis size. Only going back would be the problem.

BTW, I don't think that taking the absolute value of the difference makes sense. When the difference is zero, then obviously the estimate isn't reliable anymore and an error should be thrown.

Reply:
@artpelling, what are your thoughts?

Reply:
@artpelling, any ideas how to proceed? In case you don't care that much, I could have another shot at refactoring the module. However, I would probably only allow the number of basis/test vectors to increase to avoid the numercial issues described above.

Reply:
I have to admit that this entire PR seems quite far away at the moment. It would take me some time to get back into it (which I don't have in the next couple of weeks). Still, I would be extremely happy if the module could be refactored because I actually need to use it.

So I am very happy for you (or anyone) to take the lead on this. In May, I should have some resources to talk more about the refactoring in case there are some issues.

Reply:
Ok, will give it a shot.

Reply:
This now the changes from #2282.